### PR TITLE
[MNG-7390] Allow selecting modules outside the cwd into the reactor using --projects.

### PR DESCRIPTION
**Pull request by MartinKanters**
Simplified version of https://github.com/apache/maven/pull/659. The refactorings will be done in a separate PR for the ease of reviewing. 

There is one functional change in this PR which is described in the following ticket: MNG-7390.

---

Following this checklist to help us incorporate your
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] SUMMARY`, where you replace `MNG-XXX`
       and `SUMMARY` with the appropriate JIRA issue. Best practice is to use the JIRA issue
       title in the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by mthmulders** [comment]
Just to double-check: could the extensive suite of unit tests for this piece of code cover this new use case as well?

**Comment by MartinKanters** [comment]
> 

Tried to make that work, but it would require a huge change, since the filesystem is not really implemented/working correctly in those tests. That's why I opted for the IT.

**Review by mthmulders — APPROVED** [review]
Nice work, and good to see the change is - in fact - very simple! Once this one is merged, I'm looking forward to review the `List` -> `Set` refactoring, too.

@michael-o, did you want to review, too?

**Comment by michael-o** [comment]
Yes, tomorrow.

File: maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java [hunk 1]
```diff
@@ -138,8 +138,10 @@ public DefaultGraphBuilder( BuildResumptionDataRepository buildResumptionDataRep
     {
         ProjectDependencyGraph projectDependencyGraph = new DefaultProjectDependencyGraph( projects );
         List<MavenProject> activeProjects = projectDependencyGraph.getSortedProjects();
+        List<MavenProject> allSortedProjects = projectDependencyGraph.getSortedProjects();
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java:141]: It is the same list as above. It it correct?

**Review by michael-o — COMMENTED** [review]
I don't understand the change because `activeProjects` and `allSortedProjects` are identical.

File: maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java (hunk 1 above)
Comment (MartinKanters) [reply maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java:141]: It starts like that, but `activeProjects` is constantly being overwritten each step in the reactor the activeProjects list. This is a pattern I don't like (I prefer immutable variables). It's one of the things I fixed in the bigger refactoring.

**Comment by jira-importer** [comment]
Resolve #9175

**Pull request closed without merging**